src/NHibernate.Test/NHSpecificTest/EntityWithUserTypeCanHaveLinqGenerators/ExampleUserType.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Text;
+using NHibernate.SqlTypes;
+using NHibernate.UserTypes;
+
+namespace NHibernate.Test.NHSpecificTest.EntityWithUserTypeCanHaveLinqGenerators
+{
+    public class ExampleUserType : IUserType
+    {
+        public new bool Equals(object a, object b)
+        {
+            IExample ga = a as IExample;
+            IExample gb = b as IExample;
+            if (ga == null && gb == null)
+            {
+                return true;
+            }
+            if (ga != null && gb != null && (ga.GetType() == gb.GetType()))
+            {
+                return ga.Value == gb.Value;
+            }
+            return false;
+        }
+
+        public int GetHashCode(object x)
+        {
+            return ((IExample)x).Value.GetHashCode();
+        }
+
+        public object DeepCopy(object value)
+        {
+            if (value == null) return null;
+            if (value.GetType() == typeof(BarExample))
+            {
+                return new BarExample { Value = ((IExample)value).Value };
+            }
+            return new FooExample { Value = ((IExample)value).Value };
+        }
+
+        public object Replace(object original, object target, object owner)
+        {
+            return original;
+        }
+
+        public object Assemble(object cached, object owner)
+        {
+            return cached;
+        }
+
+        public object Disassemble(object value)
+        {
+            return value;
+        }
+
+        public SqlType[] SqlTypes { get { return new[] { SqlTypeFactory.GetStringClob(Int32.MaxValue) }; } }
+
+        public System.Type ReturnedType { get { return typeof(IExample); } }
+        public bool IsMutable { get { return true; } }
+
+        public void NullSafeSet(IDbCommand cmd, object value, int index)
+        {
+            var dataParameter = (IDataParameter)cmd.Parameters[index];
+            var example = (IExample)value;
+            dataParameter.DbType = DbType.Binary;
+            if (value == null || example.Value == null)
+            {
+                dataParameter.Value = DBNull.Value;
+            }
+            else
+            {
+                dataParameter.Value = Encoding.UTF8.GetBytes(example.ToString());
+            }
+        }
+
+        public object NullSafeGet(IDataReader rs, string[] names, object owner)
+        {
+            var index = rs.GetOrdinal(names[0]);
+            if (rs.IsDBNull(index))
+            {
+                return null;
+            }
+            var buff = new byte[1024 * 8];
+            var len = (rs).GetBytes(index, 0, buff, 0, buff.Length);
+            var val = Encoding.UTF8.GetString(buff, 0, (int)len);
+            var parts = val.Split(':');
+            if (parts[0] == "Bar")
```

Comment:
Do we really need all these technicalities in the user type to expose the issue, or can it be minimized?